server/app/services/program/ProgramServiceImpl.java

```diff
@@ -708,6 +710,51 @@ public ProgramDefinition setProgramQuestionDefinitionOptionality(
     }
   }
 
+  @Override
+  @Transactional
+  public ProgramDefinition setProgramQuestionDefinitionPosition(
+      long programId, long blockDefinitionId, long questionDefinitionId, int newPosition)
+      throws ProgramNotFoundException, ProgramBlockDefinitionNotFoundException,
+          ProgramQuestionDefinitionNotFoundException, InvalidQuestionPositionException {
+    ProgramDefinition programDefinition = getProgramDefinition(programId);
+    BlockDefinition blockDefinition = programDefinition.getBlockDefinition(blockDefinitionId);
+
+    ArrayList<ProgramQuestionDefinition> questions =
+        new ArrayList<>(blockDefinition.programQuestionDefinitions());
+
+    if (newPosition < 0 || newPosition >= questions.size()) {
+      throw new InvalidQuestionPositionException(newPosition, questions.size());
+    }
+
+    // move question to the new position
+    ProgramQuestionDefinition question = null;
+    for (Iterator<ProgramQuestionDefinition> it = questions.iterator(); it.hasNext(); ) {
+      question = it.next();
+      if (question.id() == questionDefinitionId) {
+        it.remove();
+        break;
+      }
+    }
+    if (question == null) {
+      throw new ProgramQuestionDefinitionNotFoundException(
+          programId, blockDefinitionId, questionDefinitionId);
+    }
+    questions.add(newPosition, question);
```

Comment:
Would the following be equivalent (though technically less performant due to the double iteration):
```
Optional<Question> toMove = questions.stream()
  .filter(q -> q.id() == questionDefinitionId)
  .findFirst();
List<Question> otherQuestions = questions.stream()
  .filter(q -> q.id() != questionDefinitionId)
  .collect(Collectors.toList());
otherQuestions.add(newPosition, question);

```

Reply:
Sure. I don't think performance here is critical given that number of questions should be low.